### PR TITLE
zoom at cursor

### PR DESCRIPTION
**Pull request by i-make-robots**

### CODE REVIEW

**Comment by coliss86** [comment]
When the app starts, the zoom is far away : 
image

The zoom is inconsistent: when zooming on the center, it works OK but on the edge it seems limited.
Have a look on the attached video

https://user-images.githubusercontent.com/23615562/161346562-80d21acb-64c5-4077-a993-33bc7f4f6c4b.mov

File: src/main/java/com/marginallyclever/makelangelo/preview/Camera.java [hunk 1]
```diff
@@ -27,37 +28,57 @@ public Camera() {}
...
+	 */
+	public void zoom(int amount, Point2D cursor) {
+		Point2D before = screenToWorldSpace(cursor);
+		zoom -= (double)amount * ZOOM_STEP_SIZE;
 		limitCameraZoom();
+		Point2D after = screenToWorldSpace(cursor);
+
+		offsetX -= after.x - before.x;
+		offsetY -= after.y - before.y;
+		//System.out.println("zoom="+zoom+" @ "+cursor.x+","+cursor.y);
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/preview/Camera.java:72]: please remove commented code

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java [hunk 2]
```diff
@@ -152,10 +174,12 @@ public void reshape(GLAutoDrawable glautodrawable, int x, int y, int width, int
 
 		gl2.glMatrixMode(GL2.GL_PROJECTION);
 		gl2.glLoadIdentity();
+		glu.gluOrtho2D(-width/2, width/2, -height/2, height/2);
+		/*
 		glu.gluPerspective( 90,
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java:179]: please remove commented code

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java [hunk 3]
```diff
@@ -221,16 +240,35 @@ public void display(GLAutoDrawable glautodrawable) {
...
-		
-		gl2.glLineWidth((float)camera.getZoom());
 
 		for( PreviewListener p : previewListeners ) {
 			gl2.glPushMatrix();
 			p.render(gl2);
 			gl2.glPopMatrix();
 		}
+
+		//paintCursor(gl2);
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java:250]: please remove commented code

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java [hunk 4]
```diff
@@ -221,16 +240,35 @@ public void display(GLAutoDrawable glautodrawable) {
...
 	}
-	
+
+	private void paintCursor(GL2 gl2) {
+		gl2.glPushMatrix();
+			Rectangle r = this.getBounds();
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java:255]: Please adjust the identation

**Review by coliss86 — CHANGES_REQUESTED** [review]
The feature is very nice, I didn't know how to implement it !

File: src/main/java/com/marginallyclever/makelangelo/preview/Camera.java (hunk 1 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/preview/Camera.java:72]: okay

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java (hunk 4 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java:255]: I use the indentation to very clearly show every push has a pop.  ....I'll do it.

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java (hunk 3 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java:250]: added a comment to explain what it is for.  future me will be glad it is there.

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java (hunk 2 above)
Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java:179]: Personally I would have a hard time remember how to switch back to perspective mode, this comment may be useful in future.

**Comment by i-make-robots** [comment]
wtf... why did it list cherry picked commits here.

File: src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java (hunk 2 above)
Comment (coliss86) [reply src/main/java/com/marginallyclever/makelangelo/preview/PreviewPanel.java]: git is the memory of the code that is why you have to create some meaningful commits messages in order to find old code. Comments are used to explain why how the code works if it is uncleared.
In intellij, right click on your file > Git > Show history and you'll find all the modifications done on this file.

**Comment by i-make-robots** [comment]
the one that got away is the `repaint();` in `mouseMoved()`.  it would have redrawn the picture every time the mouse moved, or 100% cpu all the time.

**Pull request merged**